improver/cli/generate_solar_time.py

```diff
@@ -0,0 +1,58 @@
+# -*- coding: utf-8 -*-
+# -----------------------------------------------------------------------------
+# (C) British Crown Copyright 2017-2021 Met Office.
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# * Redistributions of source code must retain the above copyright notice, this
+#   list of conditions and the following disclaimer.
+#
+# * Redistributions in binary form must reproduce the above copyright notice,
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""Script to run GenerateSolarTime ancillary generation."""
+
+from improver import cli
+
+
+@cli.clizefy
+@cli.with_output
+def process(target_grid: cli.inputcube, time: cli.inputdatetime):
+    """Generate a cube containing local solar time, evaluated on the target grid for
+    specified time. Local solar time is used as an input to the RainForests calibration
+    for rainfall.
+
+    Args:
+        target_grid (iris.cube.Cube):
+            A cube with the desired grid.
+        time (str):
+            A datetime specified in the format YYYYMMDDTHHMMZ at which to calculate the
+            local solar time.
+
+    Returns:
+        iris.cube.Cube:
+            A cube containing local solar time.
+    """
+    from improver.generate_ancillaries.generate_derived_solar_fields import (
+        GenerateSolarTime,
```

Comment:
Thanks @bayliffe the comment. The implementation we have for solar time (in [raincal](https://github.com/metoppv/raincal/blob/37a54422e173875d848cd3a59f25e68b0d9dbc29/src/raincal/ancil/solar_time.py#L23)) is based directly on the functionality in https://github.com/metoppv/improver/blob/master/improver/utilities/solar.py. 

The relevant function in question is [calc_solar_hour_angle](https://github.com/metoppv/improver/blob/51644de4602b0b15237a19b3ed900d165038c10d/improver/utilities/solar.py#L73). As this function allows for the evaluation of solar hour angle using ndarrays, we have been able to use this for evaluation of solar time over the grid.

To output solar time we will need to make some minor changes. Options here include:
* splitting this function to one that evaluates solar time and another that evaluates solar hour angle (using solar time), or
* leave this function alone and have another evaluate solar time from solar hour angle.

I'm open to other suggestions if you can think of a better way to handle this.

Another change we will put forwards is to add functions: `get_day_of_year` and `get_utc_hour` using the method described in `DayNightMask` process function; this will allow for reuse of existing functionality in solar.py for use in our irradiance calculations.

We've tried to reuse existing improver code as much as possible, and make changes where it makes sense to do so.

Reply:
Thanks @benowen-bom, I'm happy with whichever approach suits you guys best and minimises replication of code. I hadn't previously seen the underlying raincal code to appreciate you were basing this on the stuff in solar.py.